### PR TITLE
DropDownEditor: Set correct aria-autocomplete

### PR DESCRIPTION
**Pull request by marker-dao**

### CODE REVIEW

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.markup.tests.js [hunk 1]
```diff
@@ -114,23 +113,6 @@ module('DropDownEditor markup', {
     });
 });
 
-
-module('aria accessibility', () => {
-    test('aria role', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.markup.tests.js:119]: Seems like we missed this test for role now.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.tests.js [hunk 2]
```diff
@@ -2140,4 +2135,68 @@ QUnit.module('aria accessibility', () => {
...
+        const defaultConfigurationTrue = {
+            acceptCustomValue: true,
+            searchEnabled: true,
+        };
+
+        const configurations = [
+            [{
+                acceptCustomValue: true,
+                searchEnabled: false,
+            }, 'list'],
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.tests.js]: when `searchEnabled=false` no list is shown on typing. Are u sure we should have "list" value here?
Let's discuss in personal.

File: packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js [hunk 3]
```diff
@@ -487,11 +487,8 @@ const DropDownList = DropDownEditor.inherit({
         this.setAria('owns', this._popup && this._popupContentId);
     },
 
-    _setDefaultAria: function() {
-        this.setAria({
-            'haspopup': 'listbox',
-            'autocomplete': 'list'
-        });
+    _getHaspopupAria() {
```
Comment (ksercs) [inline packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js]:
```suggestion
    _getHasPopupAria() {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.tests.js [hunk 4]
```diff
@@ -2140,4 +2135,68 @@ QUnit.module('aria accessibility', () => {
             });
         });
     });
+
+    QUnit.module('aria-autocomplete', () => {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.tests.js]: actually file with "markup" tests is a more correct place for such checks

File: packages/devextreme/js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 5]
```diff
@@ -283,10 +283,25 @@ const DropDownEditor = TextBox.inherit({
...
+
+        return autocompleteAria;
+    },
+
     _setDefaultAria: function() {
         this.setAria({
-            'haspopup': 'true',
-            'autocomplete': 'list'
+            'haspopup': this._getHaspopupAria(),
+            'autocomplete': this._getAutocompleteAria(),
```
Comment (ksercs) [inline packages/devextreme/js/ui/drop_down_editor/ui.drop_down_editor.js]: What about these properties runtime change?
Add tests please.

File: packages/devextreme/js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 6]
```diff
@@ -283,10 +283,25 @@ const DropDownEditor = TextBox.inherit({
...
+    _getAutocompleteAria() {
+        const { searchEnabled, acceptCustomValue, readOnly, disabled } = this.option();
+
+        const isEditable = acceptCustomValue || searchEnabled;
+        const isDisabled = readOnly || disabled;
+
+        const autocompleteAria = isEditable && !isDisabled ? 'list' : 'none';
+
+        return autocompleteAria;
+    },
```
Comment (ksercs) [inline packages/devextreme/js/ui/drop_down_editor/ui.drop_down_editor.js:292]: We also need to consider a scenario in SelectBox when input substitution is rendered on typing. I believe this attr should be different in this case. 
You can look for `_shouldSubstitutionBeRendered` method in SelectBox.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.markup.tests.js (hunk 1 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.markup.tests.js:119]: No, this test already in the dropDownEditor.tests.js

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.tests.js (hunk 4 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.tests.js]: [0ff4d1d3afcafe181316848c94f2963d6e7e6cba]

File: packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js (hunk 3 above)
Comment (marker-dao) [reply packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js]: [d55bd8007ab21c760963e673ac4afd7ba711fbe7]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.markup.tests.js (hunk 1 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dropDownEditor.markup.tests.js:119]: Moved all of ally tests into markup file: [0ff4d1d3afcafe181316848c94f2963d6e7e6cba].

File: packages/devextreme/js/ui/drop_down_editor/ui.drop_down_editor.js (hunk 5 above)
Comment (marker-dao) [reply packages/devextreme/js/ui/drop_down_editor/ui.drop_down_editor.js]: [504ae75850d50b94163766a85f7f12e310748de1]

File: packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js [hunk 7]
```diff
@@ -487,11 +487,8 @@ const DropDownList = DropDownEditor.inherit({
         this.setAria('owns', this._popup && this._popupContentId);
     },
 
-    _setDefaultAria: function() {
-        this.setAria({
-            'haspopup': 'listbox',
-            'autocomplete': 'list'
-        });
+    _getAriaHasPopup() {
+        return 'listbox';
```
Comment (ksercs) [inline packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js:491]: I understand that u only do a refactoring here, but I see that test for this is missed. Could u please additionally add it?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js [hunk 8]
```diff
@@ -115,11 +115,32 @@ QUnit.module('widget sizing render', () => {
...
+        const instance = $element.dxAutocomplete('instance');
+
+        assert.strictEqual($input.attr('aria-autocomplete'), 'list', 'aria-autocomplete attribute is list');
+
+        instance.option({ searchMode: 'startswith' });
+
+        assert.strictEqual($input.attr('aria-autocomplete'), 'list', 'aria-autocomplete attribute is still list');
+    });
+
+    QUnit.test('aria-autocomplete attribute is none', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js]:
```suggestion
    QUnit.test('aria-autocomplete attribute is none if input is not editable', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js [hunk 9]
```diff
@@ -115,11 +115,32 @@ QUnit.module('widget sizing render', () => {
 });
 
 QUnit.module('aria accessibility', {}, () => {
-    QUnit.test('aria-autocomplete property', function(assert) {
+    QUnit.test('aria-autocomplete attribute is list', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js]:
```suggestion
    QUnit.test('aria-autocomplete attribute is list if input is editable', function(assert) {
```

File: packages/devextreme/js/__internal/ui/m_select_box.ts [hunk 10]
```diff
@@ -427,6 +427,26 @@ const SelectBox = (DropDownList as any).inherit({
...
 
+  _getAriaAutocomplete() {
+    const {
+      disabled,
+      readOnly,
+      searchEnabled,
+      searchMode,
+    } = this.option();
+
+    const isAvailable = !(readOnly || disabled);
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/m_select_box.ts]:
```suggestion
    const isInputEditable = !(readOnly || disabled);
```

File: packages/devextreme/js/__internal/ui/m_autocomplete.ts [hunk 11]
```diff
@@ -72,7 +72,14 @@ const Autocomplete = (DropDownList as any).inherit({
   _initMarkup() {
     this.callBase();
     this.$element().addClass(AUTOCOMPLETE_CLASS);
-    this.setAria('autocomplete', 'inline');
+  },
+
+  _getAriaAutocomplete() {
+    const { disabled, readOnly } = this.option();
+
+    const isAvailable = !(readOnly || disabled);
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/m_autocomplete.ts]:
```suggestion
    const isInputEditable = !(readOnly || disabled);
```

File: packages/devextreme/js/__internal/ui/m_select_box.ts [hunk 12]
```diff
@@ -427,6 +427,26 @@ const SelectBox = (DropDownList as any).inherit({
...
+    const hasAutocomplete = searchEnabled && isAvailable;
+
+    if (!hasAutocomplete) {
+      return 'none';
+    }
+
+    const autocompleteAria = searchMode === 'contains' ? 'list' : 'both';
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/m_select_box.ts]: I also see that inline autocomplete is not supported when acceptCustomValue=true.

I suggest u make a separate function `isInlineAutocompleteEnabled` checking that 
```javascript
            this.option('searchEnabled')
            && !this.option('acceptCustomValue')
            && this.option('searchMode') === 'startswith';
 ```
, use it here and inside of `_shouldSubstitutionBeRendered` to reduce copypaste.
 
 Don't forget to add a test as well.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js [hunk 13]
```diff
@@ -6040,6 +6065,20 @@ if(devices.real().deviceType === 'desktop') {
                 helper.checkAttributes($listItemContainer, {});
             });
         });
+
+        ['disabled', 'readOnly'].forEach(disabledOption => {
+            QUnit.test(`Input aria-autocomplete is none if ${disabledOption} is true`, function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]: let's also add tests checking that the attribute is "none" if some of these options is set to `true` on init?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js [hunk 14]
```diff
@@ -6014,13 +6032,20 @@ if(devices.real().deviceType === 'desktop') {
...
+
                 delete inputAttributes.readonly;
 
                 if(searchEnabled) {
                     inputAttributes.readonly = '';
                 }
 
                 helper.widget.option('searchEnabled', !searchEnabled);
+                inputAttributes['aria-autocomplete'] = getAriaAutocomplete(!searchEnabled, 'startswith');
+
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js:6048]: It's really difficult for me to get from this tests if we check all scenarios, so I gonna trust u about this.
I believe we should split this enormous test to several atomic in the future.

File: packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js (hunk 7 above)
Comment (marker-dao) [reply packages/devextreme/js/ui/drop_down_editor/ui.drop_down_list.js:491]: [5ac9c29211539761c3bfdfc0f3795855ae1ecc82]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js (hunk 8 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js]: [bf986670d0cf1f5213828717774550fdf66f9583]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js (hunk 9 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/autocomplete.markup.tests.js]: [bf986670d0cf1f5213828717774550fdf66f9583]

File: packages/devextreme/js/__internal/ui/m_select_box.ts (hunk 10 above)
Comment (marker-dao) [reply packages/devextreme/js/__internal/ui/m_select_box.ts]: [bf986670d0cf1f5213828717774550fdf66f9583]

File: packages/devextreme/js/__internal/ui/m_autocomplete.ts (hunk 11 above)
Comment (marker-dao) [reply packages/devextreme/js/__internal/ui/m_autocomplete.ts]: [bf986670d0cf1f5213828717774550fdf66f9583]

File: packages/devextreme/js/__internal/ui/m_select_box.ts (hunk 12 above)
Comment (marker-dao) [reply packages/devextreme/js/__internal/ui/m_select_box.ts]: [623a2f1e37e18a68cb03cafce4131873b3afc72a]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js (hunk 14 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js:6048]: [b670242fd1ef0f0f856680661f3802602b72d41c]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js (hunk 13 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/selectBox.tests.js]: Added

**Review by ksercs — APPROVED** [review]

File: packages/devextreme/js/__internal/ui/m_autocomplete.ts [hunk 15]
```diff
@@ -72,7 +72,14 @@ const Autocomplete = (DropDownList as any).inherit({
   _initMarkup() {
     this.callBase();
     this.$element().addClass(AUTOCOMPLETE_CLASS);
-    this.setAria('autocomplete', 'inline');
+  },
+
+  _getAriaAutocomplete() {
+    const { disabled, readOnly } = this.option();
+
```
Comment (iBat) [inline packages/devextreme/js/__internal/ui/m_autocomplete.ts:79]: Minor: unnecessary empty line

**Review by iBat — APPROVED** [review]

File: packages/devextreme/js/__internal/ui/m_autocomplete.ts (hunk 15 above)
Comment (marker-dao) [reply packages/devextreme/js/__internal/ui/m_autocomplete.ts:79]: For good reading.

**Review by alexanderPolosatov — APPROVED** [review]
